Mqe 1647 Read vault token from file system

jilu1 opened the pull request:
<!--- Provide a general summary of the Pull Request in the Title above -->

### Description
<!--- Provide a description of the changes proposed in the pull request -->

### Fixed Issues (if relevant)
<!--- Provide a list of fixed issues in the format magento/magento2-functional-testing-framework#<issue_number>, if relevant  -->
1. magento/magento2-functional-testing-framework#<issue_number>: Issue title
2. ...

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/verification tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)
 - [ ] Changes to Framework doesn't have backward incompatible changes for tests or have related Pull Request with fixes to tests

okolesnyk commented on src/Magento/FunctionalTestingFramework/DataGenerator/Handlers/SecretStorage/VaultStorage.php:
```diff
@@ -120,4 +134,82 @@ private function authenticated()
+            // If HOME is not set, don't fail right away
+            $homeDir = '~/';
+        } else {
+            $homeDir = rtrim($homeDir, '/') . '/';
```
It's unnecessary to do rtrim and then add slash afterwards. 
I do suggest to use `realpath` php function which will return you proper path to directory you trying to reach.

jilu1 replied:
That's a good idea. I will use realpath().

jilu1 commented on etc/config/.env.example:
```diff
@@ -32,7 +32,6 @@ BROWSER=chrome
 
 #*** Uncomment and set vault base url and access token if you want to use vault to manage _CREDS secrets ***#
 #CREDENTIAL_VAULT_BASE_URL=
-#CREDENTIAL_VAULT_TOKEN=
```
I don't think we should add VAULT_CONFIG_PATH in MFTF. For example, if user set this env in MFTF, we are not supposed to load it, nor it will work with vault.

jilu1 commented on src/Magento/FunctionalTestingFramework/DataGenerator/Handlers/SecretStorage/VaultStorage.php:
```diff
@@ -120,4 +134,82 @@ private function authenticated()
+     */
+    private function execVaultTokenHelper($cmd)
+    {
+        exec($cmd, $out, $status);
```
There are two cases for the getting the vault token (default file or custom script). We can use file_get_content for the first case, but for the second case, we have to run custom script. That's why I choose to use a generic solution.

okolesnyk commented on src/Magento/FunctionalTestingFramework/DataGenerator/Handlers/SecretStorage/VaultStorage.php:
```diff
@@ -21,6 +21,21 @@ class VaultStorage extends BaseStorage
+     */
+    const CONFIG_PATH_ENV_VAR = 'VAULT_CONFIG_PATH';
+
+    const TOKEN_HELPER_REGEX = "~\s*token_helper\s*=(.+)$~";
```
\s*token_helper\s*=(?<SOMETHING_YOU_LIKE>.+)$

jilu1 replied:
Done

okolesnyk commented on src/Magento/FunctionalTestingFramework/DataGenerator/Handlers/SecretStorage/VaultStorage.php:
```diff
@@ -120,4 +134,82 @@ private function authenticated()
+        $tokenHelper = '';
+        foreach ($lines as $line) {
+            preg_match(self::TOKEN_HELPER_REGEX, $line, $matches);
+            if (isset($matches[1])) {
```
$matches['SOMETHING_YOU_LIKE']

okolesnyk commented on src/Magento/FunctionalTestingFramework/DataGenerator/Handlers/SecretStorage/VaultStorage.php:
```diff
@@ -120,4 +134,82 @@ private function authenticated()
+        $vaultTokenFile = $homeDir . self::TOKEN_FILE;
+        if (file_exists($vaultTokenFile)) {
+            // Found .vault-token file in default location, construct command
+            $cmd = 'cat ' .  $vaultTokenFile;
```
`return file_get_content`

jilu1 commented:
> Please rethink how `readVaultTokenFromFileSystem` method works.
> It relying on UNIX console commands which will fail on Windows machine.
> If ENV HOME is not available read from `.env` and do not overcomplicate the logic.
> 
> I do suggest to use native PHP functions to read it.

Had discussion in person and addressed the issues we agreed upon.

okolesnyk commented:
@jilu1 please resolve merge conflicts and I think this ticket is good to go.

jilu1 commented:
> @jilu1 please resolve merge conflicts and I think this ticket is good to go.

It's based on previous develop where we have the other vault story in place. Now it's been reverted and thus the merge conflict. I think i will wait for 2.4.3 tag cut.

jilu1 commented:
Combined in PR #410
